{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "%run ../notebook-init.py\n",
    "\n",
    "# load R libraries                                                                                    \n",
    "%R invisible(library(ggplot2))\n",
    "%R invisible(library(fastcluster))\n",
    "%R invisible(library(reshape))\n",
    "%R invisible(library(reshape2))\n",
    "%R invisible(library(gplots))\n",
    "%R invisible(library(RSQLite))\n",
    "\n",
    "#set up ggplot2 defaults                                                                              \n",
    "%R theme_set(theme_gray(base_size=18))\n",
    "\n",
    "%pylab inline\n",
    "pylab.rcParams['figure.figsize'] = (20, 20)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AC-DIAMOND FAST\n",
      "rumen : (0.56098696648721968, 0.037682186497172525)\n",
      "filtered : (0.67655579737798133, 0.097130202536103938)\n",
      "cowrumen : (0.36098364658512183, 0.037225112641710564)\n",
      "gpigfecal : (0.15674018308756352, 0.021303337094552302)\n",
      "human : (0.61353927357687854, 0.11797808672311676)\n",
      "KRAKEN\n",
      "rumen : (0.074350605978206363, 0.015339479815349865)\n",
      "filtered : (0.22414082702759297, 0.090995852961512289)\n",
      "cowrumen : (0.025342301232090612, 0.0031820706302615864)\n",
      "gpigfecal : (0.037189487458815625, 0.015610710761141441)\n",
      "human : (0.26936412296020734, 0.1689764696662209)\n"
     ]
    }
   ],
   "source": [
    "# Count proportion of reads classified by AC-Diamond Fast and Kraken\n",
    "\n",
    "import gzip\n",
    "import sys\n",
    "import subprocess as sp\n",
    "import numpy as np\n",
    "import glob\n",
    "\n",
    "def countAligned(cinfile):\n",
    "    containingDir = '/'.join(cinfile.split('/')[:-1])\n",
    "    base = cinfile.split('/')[-1]\n",
    "    species = base.split('.')[1].split('-')[0].split('_')[0]\n",
    "    \n",
    "    if '.acdmndfast.' in base:\n",
    "        root = base[15:]\n",
    "        root = root[: root.index('.acdmndfast.tsv.gz')]\n",
    "    elif '.dmndfast.' in base:\n",
    "        root = base[15:]\n",
    "        root = root[: root.index('.dmndfast.tsv.gz')]\n",
    "    elif '.classified.' in base:\n",
    "        root = base[5:]\n",
    "        root = root[: root.index('.classified.tsv.gz')]\n",
    "        \n",
    "    linecountfile = containingDir + '/lines_in.' + root + '.fastq.gz'\n",
    "    try:\n",
    "        with open(cinfile) as c:\n",
    "            C = int(c.read().strip())\n",
    "        with open(linecountfile) as l:\n",
    "            L = int(l.read().strip())\n",
    "        P = C / (L / 4.0)\n",
    "        return (species, P)\n",
    "    except:\n",
    "        return False\n",
    "\n",
    "def mergeAligned(outputs):\n",
    "    merged = {}\n",
    "    for species, out in outputs:\n",
    "        if species not in merged:\n",
    "            merged[species] = []\n",
    "        merged[species].append(out)\n",
    "    return merged\n",
    "\n",
    "    \n",
    "def crunchAligned(merged):\n",
    "    crunched = {}\n",
    "    for sp, vals in merged.items():\n",
    "        u = np.mean(vals)\n",
    "        sd = np.std(vals)\n",
    "        crunched[sp] = (u,sd)\n",
    "    return crunched\n",
    "\n",
    "    \n",
    "def main(args):\n",
    "    if len(args) == 0:\n",
    "        print(\"Usage: unique-seq-id-files\")\n",
    "        return 1\n",
    "\n",
    "    outputs = []\n",
    "    for f in args:\n",
    "        a = countAligned(f)\n",
    "        if not a:\n",
    "            continue\n",
    "        outputs.append(a)\n",
    "    \n",
    "    merged = mergeAligned(outputs)\n",
    "    crunched = crunchAligned(merged)\n",
    "    for k,v in crunched.items():\n",
    "        print(\"{} : {}\".format(k,v))\n",
    "    return crunched\n",
    "\n",
    "        \n",
    "print('AC-DIAMOND FAST')\n",
    "files = glob.glob('../line-counts/unique-seq-ids*')\n",
    "main(files)\n",
    "\n",
    "print('KRAKEN')\n",
    "files = glob.glob('../line-counts/c_in*')\n",
    "foo = main(files)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                  Archaea                        Bacteria  \\\n",
      "cow        (47687.8564864, 31782.5803761)  (952297.902674, 31782.4887042)   \n",
      "filtered   (414.214000415, 176.144950445)  (999529.574194, 201.984696156)   \n",
      "gpigfecal  (5947.09360352, 2946.97458962)  (994049.721574, 2948.38205831)   \n",
      "human      (1528.03356317, 4403.62319599)  (998261.202815, 4459.28212911)   \n",
      "rumen      (5103.37955674, 2955.45063704)  (984896.831432, 4439.35560583)   \n",
      "\n",
      "                                  Viruses  \n",
      "cow         (14.2408399156, 7.7811772283)  \n",
      "filtered   (56.2118058899, 58.2316192273)  \n",
      "gpigfecal   (3.18482296343, 2.0517725018)  \n",
      "human      (210.763621407, 1013.55515629)  \n",
      "rumen      (9999.78901145, 2692.39994652)  \n"
     ]
    }
   ],
   "source": [
    "# Count the proportion of classified reads which were assigned to different kingdoms for AC-Diamond\n",
    "\n",
    "import gzip\n",
    "import sys\n",
    "import subprocess as sp\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "with open('../phylum2kingdom.txt') as p2k:\n",
    "    phylumtokingdom = p2k.read().strip()\n",
    "\n",
    "\n",
    "def countAligned(alignfile,p2k):\n",
    "    base = alignfile.split('/')[-1]\n",
    "    #species = base.split('.')[1].split('-')[0].split('_')[0]\n",
    "    species = alignfile.split('/')[-1].split('.')[0].split('-')[0].split('_')[0]\n",
    "    \n",
    "    if 'sheep' in alignfile:\n",
    "        species = 'sheep'\n",
    "    if 'cow' in alignfile:\n",
    "        species='cow'\n",
    "    sind = alignfile.index('phylum')\n",
    "    sfile = alignfile[:sind] + 'species' + alignfile[sind+6:]\n",
    "    #print(sfile)\n",
    "    \n",
    "    out = {\n",
    "        'Bacteria':0,  \n",
    "        'Viruses':0,\n",
    "        'Archaea':0,           \n",
    "    }\n",
    "    with gzip.open(alignfile) as af:\n",
    "        for line in af:\n",
    "            #print(line)\n",
    "            line = line.split()\n",
    "            if len(line) != 2 or 'taxa' in line[0]:\n",
    "                continue\n",
    "            k = p2k[line[0].strip()]\n",
    "#            print(\"{} -> {} ({})\".format(line[0].strip(),k,int(line[1])))\n",
    "            out[k] = out[k] + int(line[1])\n",
    "    try:\n",
    "        cmd = \"zcat {} | grep -i 'virus'\".format(sfile)\n",
    "        data = sp.check_output(cmd,shell=True)\n",
    "        for line in data.split('\\n'):\n",
    "            line = line.split()\n",
    "            if len(line) != 2 or 'Sample' in line[0]:\n",
    "                continue\n",
    "            k = 'Viruses'\n",
    "            out[k] = out[k] + int(line[1])\n",
    "    except:\n",
    "        pass\n",
    "    \n",
    "    out['Bacteria'] = out['Bacteria'] - out['Viruses']\n",
    "    N = sum([v for v in out.values()])\n",
    "    for k,v in out.items():\n",
    "        out[k] = 1000*1000*float(v)/N\n",
    "\n",
    "    return (species, out)\n",
    "\n",
    "\n",
    "def mergeAligned(outputs):\n",
    "    merged = {}\n",
    "    for species, out in outputs:\n",
    "        for kind, val in out.items():\n",
    "            if kind not in merged:\n",
    "                merged[kind] = {}\n",
    "            if species not in merged[kind]:\n",
    "                merged[kind][species] = []\n",
    "            merged[kind][species].append(val)\n",
    "    return merged\n",
    "\n",
    "def buffer(outputs):\n",
    "    buff = {}\n",
    "    for species, out in outputs:\n",
    "        if species not in buff:\n",
    "            buff[species] = []\n",
    "        buff[species].append(out)\n",
    "\n",
    "    for species,outs in buff.items():\n",
    "        top = 0\n",
    "        if len(outs) > len(top):\n",
    "            top = len(outs)\n",
    "        for out in outs:\n",
    "            for _ in range(top - len(out)):\n",
    "                out.append(0)\n",
    "    \n",
    "\n",
    "def crunchAligned(merged):\n",
    "    crunched = {kind:{} for kind in merged.keys()}\n",
    "    for kind, val in merged.items():\n",
    "        for species, vals in val.items():\n",
    "            u = np.mean(vals)\n",
    "            sd = np.std(vals)\n",
    "            crunched[kind][species] = (u,sd)\n",
    "    return crunched\n",
    "\n",
    "def prettyprint(crunched):\n",
    "    print(\"\\n--------------------------------------------------------\\n\")\n",
    "    outline = \"{:.1f} ({:.1f}) \\t& {:.1f} ({:.1f}) \\t& {:.1f} ({:.1f}) \\t& {:.1f} ({:.1f}) \\t\\\\\\\\\"\n",
    "    for row in crunched.itertuples():\n",
    "        print(row)\n",
    "        ar = []\n",
    "        for v in row[1:]:\n",
    "  #          print(v)\n",
    "            ar.append(float(v[0]))\n",
    "            ar.append(float(v[1]))\n",
    " #       print(ar)\n",
    "        print(outline.format(*ar))\n",
    "        \n",
    "    \n",
    "def main(args):\n",
    "    if len(args) == 0:\n",
    "        print(\"Usage: metaphlan-profile-files\")\n",
    "        return 1\n",
    "\n",
    "    p2k = {}\n",
    "    a = phylumtokingdom.split('\\n')\n",
    "    a = [v for v in a if len(v) > 0]\n",
    "    for i in range(0,len(a),2):\n",
    "        p2k[a[i].strip()] = a[i+1].strip()\n",
    "\n",
    "#     for k,v in p2k.items():\n",
    "#         print(\"{} : {}\".format(k,v))\n",
    "    \n",
    "    outputs = []\n",
    "    for f in args:\n",
    "        a = countAligned(f,p2k)\n",
    "        if not a:\n",
    "            continue\n",
    "        outputs.append(a)\n",
    "    \n",
    "    merged = mergeAligned(outputs)\n",
    "    crunched = crunchAligned(merged)\n",
    "    crunched = pd.DataFrame(crunched)\n",
    "    print(crunched)\n",
    "    #prettyprint(crunched)\n",
    "\n",
    "files = glob.glob('../taxonomic-profiles/ac-diamond-fast/individual-profiles/*phylum*')\n",
    "main(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../line-counts/kraken_phyla.human-posterior_fornix-R056796.classified.tsv\n",
      "                                  Archaea                        Bacteria  \\\n",
      "cowrumen   (62985.3027155, 30345.0069695)  (869397.718672, 34322.4963739)   \n",
      "filtered   (368.994187329, 226.225373368)  (998713.041574, 475.982607408)   \n",
      "gpigfecal  (6222.03133467, 3980.41843104)   (993118.46342, 4011.76636231)   \n",
      "human      (5590.04867586, 17023.0934813)   (982380.95938, 26917.3489289)   \n",
      "rumen      (13263.0803366, 6845.71852958)  (917390.322527, 23109.2876915)   \n",
      "setA1                (14831.0834705, 0.0)            (985087.959438, 0.0)   \n",
      "setB1                (69094.8236298, 0.0)            (930822.374725, 0.0)   \n",
      "setB2                 (69088.446011, 0.0)            (930835.866591, 0.0)   \n",
      "\n",
      "                                  Viruses  \n",
      "cowrumen   (67616.9786129, 9585.21486643)  \n",
      "filtered   (917.964239157, 309.533142928)  \n",
      "gpigfecal  (659.505245315, 214.960768491)  \n",
      "human      (12028.9919443, 23526.7720458)  \n",
      "rumen      (69346.5971367, 20578.9959433)  \n",
      "setA1                 (80.957091856, 0.0)  \n",
      "setB1                (82.8016451258, 0.0)  \n",
      "setB2                (75.6873983164, 0.0)  \n"
     ]
    }
   ],
   "source": [
    "# Count the proportion of classified reads which were assigned to different kingdoms for AC-Diamond\n",
    "\n",
    "import gzip\n",
    "import sys\n",
    "import subprocess as sp\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def countAligned(alignfile):\n",
    "    base = alignfile.split('/')[-1]\n",
    "    tool = 'kraken'\n",
    "    species = alignfile.split('/')[-1].split('.')[1].split('-')[0].split('_')[0]\n",
    "    \n",
    "    \n",
    "    out = {\n",
    "        'Bacteria':0,  \n",
    "        'Viruses':0,\n",
    "        'Archaea':0,           \n",
    "    }\n",
    "    with open(alignfile) as af:\n",
    "        for line in af:\n",
    "            line = line.split()\n",
    "            if len(line) != 2 or 'taxa' in line[0]:\n",
    "                continue\n",
    "            k = line[0].strip().split('__')[1]\n",
    "            out[k] = out[k] + int(line[1])\n",
    "    \n",
    "    N = sum([v for v in out.values()])\n",
    "    if N == 0 :\n",
    "        print(alignfile)\n",
    "        return False\n",
    "    for k,v in out.items():\n",
    "        out[k] = 1000*1000*float(v)/N\n",
    "\n",
    "    return (species, out)\n",
    "\n",
    "\n",
    "def mergeAligned(outputs):\n",
    "    merged = {}\n",
    "    for species, out in outputs:\n",
    "        for kind, val in out.items():\n",
    "            if kind not in merged:\n",
    "                merged[kind] = {}\n",
    "            if species not in merged[kind]:\n",
    "                merged[kind][species] = []\n",
    "            merged[kind][species].append(val)\n",
    "    return merged\n",
    "    \n",
    "\n",
    "def crunchAligned(merged):\n",
    "    crunched = {kind:{} for kind in merged.keys()}\n",
    "    for kind, val in merged.items():\n",
    "        for species, vals in val.items():\n",
    "            u = np.mean(vals)\n",
    "            sd = np.std(vals)\n",
    "            crunched[kind][species] = (u,sd)\n",
    "    return crunched\n",
    "        \n",
    "    \n",
    "def main(args):\n",
    "\n",
    "    outputs = []\n",
    "    for f in args:\n",
    "        a = countAligned(f)\n",
    "        if not a:\n",
    "            continue\n",
    "        outputs.append(a)\n",
    "    \n",
    "    merged = mergeAligned(outputs)\n",
    "    crunched = crunchAligned(merged)\n",
    "    crunched = pd.DataFrame(crunched)\n",
    "    print(crunched)\n",
    "\n",
    "files = glob.glob('../line-counts/kraken_phyla*')\n",
    "main(files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
